crates/ntx-builder/src/builder/server/state.rs

```diff
@@ -0,0 +1,123 @@
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+
+use miden_node_utils::network_note::NetworkNote;
+use miden_objects::{
+    note::{NoteId, NoteTag, Nullifier},
+    transaction::TransactionId,
+};
+
+/// Max number of notes taken for executing a single network transaction
+const MAX_BATCH: usize = 50;
+
+/// Maintains state of the network notes for the transaction builder.
+///
+/// Note tags are mainly kept in a [`VecDeque`], but also indexed by nullifiers and note tags in
+/// order to enable simple lookups when discarding by nullifiers and deciding which notes to
+/// execute against a unique account, respectively.
+/// Additionally, a list of inflight notes is kept to track their lifecycle.
+#[derive(Debug)]
+pub struct PendingNotes {
+    /// Contains a queue that provides ordering of accounts to execute against
+    account_queue: VecDeque<NoteTag>,
+    /// Pending network notes that have not been consumed as part of a committed transaction.
+    by_tag: BTreeMap<NoteTag, Vec<NetworkNote>>,
+    /// A map of nullifiers mapped to their note IDs
+    by_nullifier: BTreeMap<Nullifier, NoteId>,
+    /// Inflight network notes with their associated transaction IDs
+    inflight_txs: BTreeMap<TransactionId, Vec<NetworkNote>>,
+}
+
+impl PendingNotes {
+    pub fn new(unconsumed_network_notes: Vec<NetworkNote>) -> Self {
+        let mut state = Self {
+            account_queue: VecDeque::new(),
+            by_tag: BTreeMap::new(),
+            by_nullifier: BTreeMap::new(),
+            inflight_txs: BTreeMap::new(),
+        };
+        state.queue_unconsumed_notes(unconsumed_network_notes);
+        state
+    }
+
+    /// Add network notes to the pending notes queue.
+    pub fn queue_unconsumed_notes(&mut self, notes: impl IntoIterator<Item = NetworkNote>) {
+        for n in notes {
+            let tag = n.metadata().tag();
+
+            self.by_nullifier.insert(n.nullifier(), n.id());
+            self.by_tag.entry(tag).or_default().push(n.clone());
+
+            self.account_queue.push_back(tag);
```

Comment:
In case there are multiple notes per tag, we'll get duplicates in the `account_queue`, right? My thinking was that `account_queue` would contain unique tags. So, here, before adding the tag we'd check if the tags already in the queue, and if so, won't add it again.

Reply:
My thought was that having duplicate tags would be fine since you could take until one had notes remaining, but I modified it to only have one copy